subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/SpecificInstallationToolchainSpec.java

```diff
@@ -17,18 +17,53 @@
 package org.gradle.jvm.toolchain.internal;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import org.gradle.api.model.ObjectFactory;
 
 import java.io.File;
 
 public class SpecificInstallationToolchainSpec extends DefaultToolchainSpec {
 
+    public static class Key implements JavaToolchainSpecInternal.Key {
+
+        private final File javaHome;
+
+        public Key(File javaHome) {
+            this.javaHome = javaHome;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Key key = (Key) o;
+            return java.util.Objects.equals(javaHome, key.javaHome);
+        }
+
+        @Override
+        public int hashCode() {
+            return java.util.Objects.hash(javaHome);
+        }
+    }
+
     private final File javaHome;
 
     public SpecificInstallationToolchainSpec(ObjectFactory factory, File javaHome) {
         super(factory);
         this.javaHome = javaHome;
+
+        // disallow changing property values
+        getLanguageVersion().finalizeValue();
+        getVendor().finalizeValue();
+        getImplementation().finalizeValue();
```

Comment:
Similarly to `CurrentJvmToolchainSpec` changing any of the properties of the specific toolchain could also mess up caching.

It is further unclear why this class even extends `DefaultToolchainSpec`. There is more entangled code that relies on this, and changing it is not straight-forward. Still, I think this should be reworked during toolchain conventions change in 8.0.